### PR TITLE
Add reactive MS SQL client extension

### PR DESCRIPTION
**Pull request by tsegismont**
- runtime/deployment modules
- integration tests (including native mode)
- datasources docs

### CODE REVIEW

**Comment by tsegismont** [comment]
cc @Sanne @DavideD

File: extensions/reactive-mssql-client/deployment/pom.xml [hunk 1]
```diff
@@ -0,0 +1,200 @@
...
+
+    <parent>
+        <groupId>io.quarkus</groupId>
+        <artifactId>quarkus-reactive-mssql-client-parent</artifactId>
+        <version>999-SNAPSHOT</version>
+    </parent>
+
+    <artifactId>quarkus-reactive-mssql-client-deployment</artifactId>
+
+    <name>Quarkus - Reactive MS SQL Client - Deployment</name>
```
Comment (cescoffier) [inline extensions/reactive-mssql-client/deployment/pom.xml:15]: Same question about the name.
It looks from their page that the name is just "SQL Server" but SQL Server Client is weird... SQL Client is not exact. I understand how you ended up with MS SQL Client. I would still expand MS.

File: extensions/reactive-mssql-client/deployment/src/test/java/io/quarkus/reactive/mssql/client/DevModeResource.java [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+    @Path("/error")
+    @Produces(MediaType.TEXT_PLAIN)
+    public CompletionStage<Response> getErrorMessage() throws SQLException {
+        CompletableFuture<Response> future = new CompletableFuture<>();
+        client.query("SELECT 1").execute(ar -> {
+            Class<?> expectedExceptionClass = ConnectException.class;
+            if (ar.succeeded()) {
+                future.complete(Response.serverError().entity("Expected SQL query to fail").build());
+            } else if (!expectedExceptionClass.isAssignableFrom(ar.cause().getClass())) {
+                ar.cause().printStackTrace();
```
Comment (cescoffier) [inline extensions/reactive-mssql-client/deployment/src/test/java/io/quarkus/reactive/mssql/client/DevModeResource.java:33]: should be removed.

**Review by cescoffier — CHANGES_REQUESTED** [review]

File: docs/src/main/asciidoc/datasource.adoc [hunk 3]
```diff
@@ -662,6 +662,10 @@ include::{generated-dir}/config/quarkus-reactive-db2-client.adoc[opts=optional,
 
 include::{generated-dir}/config/quarkus-reactive-mysql-client.adoc[opts=optional, leveloffset=+1]
 
+=== Reactive MS SQL Specific Configuration
```
[the remark this answers is not in the document]
Comment (gsmet) [reply docs/src/main/asciidoc/datasource.adoc:665]: It's Microsoft SQL Server.

File: extensions/reactive-mssql-client/deployment/pom.xml (hunk 1 above)
Comment (Sanne) [reply extensions/reactive-mssql-client/deployment/pom.xml:15]: +1 to expand here and in the description for sake o clarity.

In other areas the short hand is fine, as it's commonly referred to as MS SQL by its users (which we are addressing).

**Comment by Sanne** [comment]
@DavideD could you follow up with integration tests using MS SQL + Hibernate Reactive ?
Which I assume will need an HR release ?

**Comment by DavideD** [comment]
Do you mean integration tests for Quarkus? Sure, [we are having somes trouble with the MS SQL Client on Hibernate Reactive](https://github.com/hibernate/hibernate-reactive/pull/865#issuecomment-872419889) at the moment, but as soon as it has been fixed, I will add them

File: docs/src/main/asciidoc/datasource.adoc (hunk 3 above)
[the remark this answers is not in the document]
Comment (tsegismont) [reply docs/src/main/asciidoc/datasource.adoc]: Done

File: extensions/reactive-mssql-client/deployment/pom.xml (hunk 1 above)
Comment (tsegismont) [reply extensions/reactive-mssql-client/deployment/pom.xml]: Done

File: extensions/reactive-mssql-client/deployment/src/test/java/io/quarkus/reactive/mssql/client/DevModeResource.java (hunk 2 above)
Comment (tsegismont) [reply extensions/reactive-mssql-client/deployment/src/test/java/io/quarkus/reactive/mssql/client/DevModeResource.java]: Done

**Comment by tsegismont** [comment]
Thank you all for the review. I addressed the concerns and rebased on `main`. This allowed to add support for the TDS packet size option that comes with Vert.x 4.1.1.

**Comment by Sanne** [comment]
Looks like the failures are related:
```
2021-07-02T18:09:30.9633222Z [ERROR] Failed to execute goal io.fabric8:docker-maven-plugin:0.31.0:start (docker-start) on project quarkus-integration-test-reactive-mssql-client: I/O Error: [mcr.microsoft.com/mssql/server:2019-latest] "quarkus-test-mssql": Timeout after 10045 ms while waiting on tcp port '[/172.17.0.2:1433]' -> [Help 1]
```

**Comment by Sanne** [comment]
it's a DB bootstrap timeout:
```
2021-07-02T17:29:38.5268124Z [ERROR] DOCKER> [mcr.microsoft.com/mssql/server:2019-latest] "quarkus-test-mssql": Timeout after 10045 ms while waiting on tcp port '[/172.17.0.2:1433]'
```

**Comment by Sanne** [comment]
@tsegismont could you copy the infrastructure from the pom.xml in `integration-tests/jpa-mssql` ? It seems your pom.xml is similar but doesn't have crucial elements like the timeout settings.

**Review by cescoffier — APPROVED** [review]
It looks good to me. However, validation by Sanne would be great (as I don't know much about databases anyway).

**Comment by tsegismont** [comment]
>@tsegismont could you copy the infrastructure from the pom.xml in integration-tests/jpa-mssql ? It seems your pom.xml is similar but doesn't have crucial elements like the timeout settings.

@Sanne good idea, I will do it for both the integration tests and the deployment module tests.

File: extensions/reactive-mssql-client/deployment/pom.xml [hunk 4]
```diff
@@ -143,13 +140,9 @@
...
-                                            <!-- good docs found at: http://dmp.fabric8.io/#build-healthcheck -->
-                                            <tcp>
-                                                <mode>direct</mode>
-                                                <ports>
-                                                    <port>1433</port>
-                                                </ports>
-                                            </tcp>
+                                            <log>SQL Server is now ready for client connections</log>
+                                            <!-- Unfortunately booting this is slow, needs to set a generous timeout: -->
+                                            <time>40000</time>
```
Comment (Sanne) [inline extensions/reactive-mssql-client/deployment/pom.xml:145]: :+1:  thanks! Let's see if CI managed to run it now.

**Review by Sanne — APPROVED** [review]

File: extensions/reactive-mssql-client/deployment/pom.xml (hunk 4 above)
Comment (tsegismont) [reply extensions/reactive-mssql-client/deployment/pom.xml:145]: There is one failure but it looks unrelated

**Comment by Sanne** [comment]
awesome, let's merge this. Many thanks!

**Pull request merged**